### PR TITLE
Added check_dtype in topologyattr.py

### LINKED ISSUE #2764: Setting attributes should have dtype checks
Our Topology system is essentially statically typed.  I recently had a test fail when you tried setting an atom name to an integer:  https://github.com/MDAnalysis/mdanalysis/pull/2755/commits/97e1782942a8262acfab2990460a129a15ed52bd

This would have previously worked as `dtype=object` (what we use for numpy arrays of python strings) will allow a Python integer into the array.  

We should have a `check_dtype` decorator around `TopologyAttr.set_atoms` (and residue/segment) to check that the supplied values are the correct type.

### PR DESCRIPTION
**Pull request by Muhammad-Ahsan-Rasheed**
Fixes #2764

Changes made in this Pull Request:
 - Added decorator for checking the dtype and set the required value to str.

PR Checklist
------------
 - [ Not Yet ] Tests?
 - [ No ] Docs?
 - [ No ] CHANGELOG updated?
 - [ 2764 ] Issue raised/referenced?

----
:books: Documentation preview :books:: https://readthedocs-preview--4113.org.readthedocs.build/en/4113/

### CODE REVIEW

**Comment by richardjgowers** [comment]
@Muhammad-Ahsan-Rasheed can you address the comments of the bots please?

File: .gitignore [hunk 1]
```diff
@@ -52,3 +52,6 @@ benchmarks/results
 .idea
 .vscode
 *.lock
+
+# add mdanalysis dir 
+mdanalysis/
```
Comment (RMeli) [inline .gitignore]: Why was this added? If this is only specific to your setup, please remove it.

Please also remove all the files (`.dll`, `.exe`, ...) that have been added to this PR from such folder.

**Review by RMeli — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (Muhammad-Ahsan-Rasheed) [reply .gitignore]: I removed the environment files, sorry about that.

**Comment by Muhammad-Ahsan-Rasheed** [comment]
> darker-main-code

I checked the status all are succeeded.

image

**Comment by orbeckst** [comment]
Your tests are failing, as can bee seen in checks. Check how these failures relate to the changes that you made.

You can also run the tests locally with `pytest` to get at the failures more quickly.

One example is
```
[gw1] linux -- Python 3.8.16 /usr/share/miniconda3/envs/test/bin/python

self = <MDAnalysisTests.coordinates.test_gro.TestGrowriterReindex object at 0x7f9debc6d310>

        @pytest.fixture()
        def u(self):
            gro = '''test
    1
        2CL      CL20850   0.000   0.000   0.000
    7.29748 7.66094 9.82962'''
            u = mda.Universe(StringIO(gro), format='gro')
>           u.atoms[0].id = 3

/home/runner/work/mdanalysis/mdanalysis/testsuite/MDAnalysisTests/coordinates/test_gro.py:441: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/runner/work/mdanalysis/mdanalysis/package/MDAnalysis/core/groups.py:294: in __setattr__
    super(_TopologyAttrContainer, self).__setattr__(attr, value)
/home/runner/work/mdanalysis/mdanalysis/package/MDAnalysis/core/groups.py:257: in setter
    return attr.__setitem__(self, values)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<MDAnalysis.core.topologyattrs.Atomids object at 0x7f9de25f0130>, <Atom 1: CL of type CL of resname CL, resid 2 and segid SYSTEM>, 3)
kwargs = {}
values = <Atom 1: CL of type CL of resname CL, resid 2 and segid SYSTEM>

    def wrapper(*args, **kwargs):
        values = kwargs.get('values', args[1])
        if not isinstance(values, dtype):
>           raise ValueError(f"Expected {dtype}, but got {type(values)}")
E           ValueError: Expected <class 'str'>, but got <class 'MDAnalysis.core.groups.Atom'>
```

**Comment by IAlibay** [comment]
@Muhammad-Ahsan-Rasheed are you still interested in working on this? If not we will go ahead and close this PR.

**Comment by IAlibay** [comment]
I'm closing this due to lack of activity, please re-open if you would like to continue this PR.
